Bio/PDB/rotamers.py

```diff
@@ -0,0 +1,279 @@
+import numpy as np
+import os
+import gzip
+
+
+DATA_DIR = os.path.join(os.path.dirname(os.path.realpath(__file__)), "rotamers")
+DUNBRACK_FILTERED = f"{DATA_DIR}/dunbrack_reduced.gz"
+
+
+class Rotamer:
+    """
+    Object to store rotamers
+    """
+
+    def __init__(self, resname, psi, phi, chi1, chi2, chi3, chi4, probability):
+        """
+        Init the rotamer with given diherdral angles
+        """
+        self.psi = float(psi)
+        self.phi = float(phi)
+        self.chi1 = float(chi1)
+        self.chi2 = float(chi2)
+        self.chi3 = float(chi3)
+        self.chi4 = float(chi4)
+        self.resname = resname
+        self.probability = float(probability)
+        self.sample_residue = False
+
+    def __getitem__(self, item):
+        """
+        Returns the given angle property of the rotamer
+        """
+        return getattr(self, item.lower())
+
+    def set_sample_residue(self, sample_residue):
+        """
+        Sets a sample residue for the rotamer object
+        This function is only called during RotamerSampling to save memory
+        """
+        self.sample_residue = sample_residue
+
+    def get_sample_residue(self):
+        """
+        Returns the stored sample residue
+        """
+        if self.sample_residue:
+            return self.sample_residue
+        raise ValueError("Sample residue is not set, use RotamerSampling to set it!")
+
+
+class RotamerSampling:
+    """
+    Object to sample rotamers based on certain criterion to use for mutation
+    """
+
+    def __init__(self, structure_object, rotamer_loc=DUNBRACK_FILTERED):
+        """
+        Load the Dunbrack rotamer library
+        :param rotamer_loc: Location of the rotamer file to use
+        :returns : Nested defaultdict of the torsion angles for each psi phi angle in the file
+        """
+        self.structure_object = structure_object
+        self._rot_lib = defaultdict(lambda: defaultdict(lambda: defaultdict(list)))
+        with gzip.open(rotamer_loc, "rt") as fn:
+            for line in fn:
+                if line.startswith("#"):
+                    continue
+                amino_acid, angle1, angle2, prob, chi1, chi2, chi3, chi4 = line.split()
+                rotamer = Rotamer(
+                    amino_acid, angle1, angle2, chi1, chi2, chi3, chi4, prob
+                )
+                self._rot_lib[amino_acid][int(angle1)][int(angle2)].append(rotamer)
+
+    def select_based_on_clashes(
+        self,
+        residue,
+        mutate_to,
+        sample_residue,
+        rotamers,
+        skip_own_chain=False,
+    ):
+        """
+        Calculates the best rotamer based on VdW energy
+        :param residue: Bio.PDB Residue object
+        :param mutate_to: Three letter amino acid code in uppercase letters
+        :param rotamers: Rotamers to use for selection. Obtained from rotamer_lib
+        :param skip_own_chain: Boolean to skip the parent chain of the selected residue. Default: False
```

Comment:
"Ignore clashes within the chain of the modified residue"?  Not sure what good wording here is (I am poor at clarity) but I struggled to understand what this meant.